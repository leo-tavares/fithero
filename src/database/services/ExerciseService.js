/* @flow */

import realm from '../index';
import type { AddExerciseType, ExerciseSchemaType } from '../types';
import type { RealmResults } from '../../types';
import { EXERCISE_SCHEMA_NAME } from '../schemas/ExerciseSchema';
import { WORKOUT_EXERCISE_SCHEMA_NAME } from '../schemas/WorkoutExerciseSchema';
import { WORKOUT_SET_SCHEMA_NAME } from '../schemas/WorkoutSetSchema';
import { WORKOUT_SCHEMA_NAME } from '../schemas/WorkoutSchema';

export const userExerciseIdPrefix = 'user-exercise--';

export const isCustomExercise = (id: string) =>
  id.includes(userExerciseIdPrefix);

const _generateId = () => {
  const exercises = realm.objects(EXERCISE_SCHEMA_NAME).sorted('id', true);
  let max = 1;
  if (exercises.length > 0) {
    max = parseInt(exercises[0].id.split(userExerciseIdPrefix)[1], 10) + 1;
  }
  // The user can create a max of 100000 exercises :)
  return `${userExerciseIdPrefix}${max.toString().padStart(6, '0')}`;
};

export const addExercise = (exercise: AddExerciseType) => {
  realm.write(() => {
    const id = _generateId();

    realm.create(EXERCISE_SCHEMA_NAME, {
      id,
      ...exercise,
    });
  });
};

export const editExercise = (newExercise: ExerciseSchemaType) => {
  realm.write(() => {
    const exercise = realm.objectForPrimaryKey(
      EXERCISE_SCHEMA_NAME,
      newExercise.id
    );
    exercise.name = newExercise.name;
    exercise.notes = newExercise.notes || null;
    exercise.primary = newExercise.primary;
    exercise.secondary = newExercise.secondary;
  });
};

export const getAllExercises = (): RealmResults<ExerciseSchemaType> =>
  realm.objects(EXERCISE_SCHEMA_NAME);

export const getExerciseById = (id: string) =>
  realm.objects(EXERCISE_SCHEMA_NAME).filtered(`id = $0`, id);

export const getExerciseObjectByPrimaryId = (id: string) =>
  realm.objectForPrimaryKey(EXERCISE_SCHEMA_NAME, id);

export const deleteExercise = (id: string) => {
  realm.write(() => {
    const workoutExercises = realm
      .objects(WORKOUT_EXERCISE_SCHEMA_NAME)
      .filtered(`type = $0`, id);

    const workoutExerciseSets = realm
      .objects(WORKOUT_SET_SCHEMA_NAME)
      .filtered(`type = $0`, id);

    // We first need to delete any exercise referenced that is part of a workout
    realm.delete([...workoutExerciseSets, ...workoutExercises]);

    // Clear workouts if necessary
    const emptyWorkouts = realm
      .objects(WORKOUT_SCHEMA_NAME)
      .filtered('exercises.@size = 0 AND comments = null');

    realm.delete(emptyWorkouts);

    // Finally when no references, delete the exercise
    const exercise: ExerciseSchemaType = realm.objectForPrimaryKey(
      EXERCISE_SCHEMA_NAME,
      id
    );
    realm.delete(exercise);
  });
};
