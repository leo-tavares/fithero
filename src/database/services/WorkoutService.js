/* @flow */

import realm from '../index';

import type { WorkoutSchemaType } from '../types';
import type { RealmResults } from '../../types';
import {
  getFirstAndLastMonthDay,
  getFirstAndLastWeekday,
  getToday,
  toDate,
} from '../../utils/date';
import { WORKOUT_SCHEMA_NAME } from '../schemas/WorkoutSchema';

export const getAllWorkouts = (): RealmResults<WorkoutSchemaType> =>
  realm.objects(WORKOUT_SCHEMA_NAME);

export const getAllWorkoutsWithExercises = () =>
  realm.objects(WORKOUT_SCHEMA_NAME).filtered('exercises.@count > 0');

export const getAllWorkoutsWithExercisesSortedByDate = () =>
  realm
    .objects(WORKOUT_SCHEMA_NAME)
    .filtered('exercises.@count > 0')
    .sorted([['date', true]]);

export const getWorkoutsByRange = (
  start: Date,
  end: Date,
  onlyWithExercises: boolean = false
) =>
  realm
    .objects(WORKOUT_SCHEMA_NAME)
    .filtered(
      `date >= $0 AND date <= $1${
        onlyWithExercises ? ' AND exercises.@count > 0' : ''
      }`,
      start,
      end
    );

export const getWorkoutById = (id: string) =>
  realm.objects(WORKOUT_SCHEMA_NAME).filtered(`id = $0`, id);

export const getWorkoutByPrimaryKey = (id: string) =>
  realm.objectForPrimaryKey(WORKOUT_SCHEMA_NAME, id);

export const getWorkoutsThisWeek = () => {
  const [start, end] = getFirstAndLastWeekday(getToday());
  return getWorkoutsByRange(start, end, true);
};

export const getWorkoutsThisMonth = () => {
  const [start, end] = getFirstAndLastMonthDay(getToday());
  return getWorkoutsByRange(start, end, true);
};

export const getWorkoutComments = (workoutId: string) => {
  const workout = realm.objectForPrimaryKey(WORKOUT_SCHEMA_NAME, workoutId);
  if (workout) {
    return workout.comments ? workout.comments : '';
  }
  return '';
};

export const setWorkoutComments = (workoutId: string, comments: string) => {
  realm.write(() => {
    let workout = realm.objectForPrimaryKey(WORKOUT_SCHEMA_NAME, workoutId);
    if (!workout) {
      workout = realm.create(WORKOUT_SCHEMA_NAME, {
        id: workoutId,
        date: toDate(workoutId),
      });
    }
    workout.comments = comments;
  });
};

export const deleteWorkoutComments = (workoutId: string) => {
  realm.write(() => {
    const workout = realm.objectForPrimaryKey(WORKOUT_SCHEMA_NAME, workoutId);
    if (workout) {
      if (workout.exercises.length === 0) {
        realm.delete(workout);
      } else {
        workout.comments = null;
      }
    }
  });
};
