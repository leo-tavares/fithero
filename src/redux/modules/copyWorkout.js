/* @flow */

import type { WorkoutSchemaType } from '../../database/types';

export type CopyWorkoutType = {
  selectedWorkout: ?WorkoutSchemaType,
};

export const SET_SELECTED_WORKOUT = 'fithero/copyWorkout/SET_SELECTED_WORKOUT';

type State = CopyWorkoutType;
type Action = {
  type: typeof SET_SELECTED_WORKOUT,
  payload: ?WorkoutSchemaType,
};

export const initialState: State = {
  selectedWorkout: null,
};

export default function reducer(state: State = initialState, action: Action) {
  switch (action.type) {
    case SET_SELECTED_WORKOUT: {
      return {
        ...state,
        selectedWorkout: action.payload,
      };
    }
    default:
      return state;
  }
}

export const setSelectedWorkout = (payload: ?WorkoutSchemaType) => ({
  type: SET_SELECTED_WORKOUT,
  payload,
});
