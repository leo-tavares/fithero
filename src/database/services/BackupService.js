/* @flow */

import Share from 'react-native-share';
import { FileSystem } from 'react-native-unimodules';
import * as DocumentPicker from 'expo-document-picker';
import AsyncStorage from '@react-native-community/async-storage';

import realm from '../';
import { EXERCISE_SCHEMA_NAME } from '../schemas/ExerciseSchema';
import { WORKOUT_SCHEMA_NAME } from '../schemas/WorkoutSchema';
import {
  firstDayOfTheWeekToNumber,
  getCurrentLocale,
  getToday,
  setMomentFirstDayOfTheWeek,
} from '../../utils/date';
import { deserializeExercises, deserializeWorkouts } from '../utils';
import { PACKAGE_NAME, Settings } from '../../utils/constants';
import type { SettingsType } from '../../redux/modules/settings';
import { Platform, StatusBar } from 'react-native';

export const backupDatabase = async () => {
  const keys = await AsyncStorage.getAllKeys();
  const asyncStorageData = await AsyncStorage.multiGet(keys);
  let settings = {};
  asyncStorageData.forEach(([key, value]) => {
    if (key.includes(PACKAGE_NAME)) {
      settings[key] = value;
    }
  });

  // Clear old backups
  try {
    const files = await FileSystem.readDirectoryAsync(
      FileSystem.cacheDirectory
    );
    files.forEach(async file => {
      if (file.includes('fithero-backup-')) {
        await FileSystem.deleteAsync(`${FileSystem.cacheDirectory}/${file}`);
      }
    });
  } catch {
    // No file
  }

  // Workouts already contain everything about workouts exercises and workouts sets
  const workoutSchema = realm.objects(WORKOUT_SCHEMA_NAME);
  const exerciseSchema = realm.objects(EXERCISE_SCHEMA_NAME);

  const backupJSON = JSON.stringify({
    workouts: deserializeWorkouts(workoutSchema),
    exercises: deserializeExercises(exerciseSchema),
    settings,
  });

  const filePath = `${
    FileSystem.cacheDirectory
  }/fithero-backup-${getToday().format('YYYY-MM-DD')}.json`;
  await FileSystem.writeAsStringAsync(filePath, backupJSON);

  await Share.open({ url: `file://${filePath}`, type: 'text/plain' });
};

export const restoreDatabase = async (
  initSettingsAction: (settings: SettingsType) => void
) => {
  const { name, uri, type } = await DocumentPicker.getDocumentAsync({
    copyToCacheDirectory: false,
  });
  if (type === 'cancel') {
    return;
  }

  if (name.match(/fithero-backup-\d\d\d\d-\d\d-\d\d.json/g)) {
    await FileSystem.copyAsync({
      from: uri,
      to: `${FileSystem.cacheDirectory}/${name}`,
    });
    const backup = JSON.parse(
      await FileSystem.readAsStringAsync(`${FileSystem.cacheDirectory}/${name}`)
    );

    const settings = [];
    Object.keys(backup.settings).map(key => {
      settings.push([key, backup.settings[key]]);
    });
    await AsyncStorage.multiSet(settings);

    const firstDayOfTheWeek = backup.settings[Settings.firstDayOfTheWeek];
    const locale = getCurrentLocale();
    setMomentFirstDayOfTheWeek(
      locale,
      firstDayOfTheWeekToNumber(backup.settings[Settings.firstDayOfTheWeek]),
      true
    );

    // Update store after setting moment
    const appTheme = backup.settings[Settings.appTheme] || 'default';
    initSettingsAction({
      appTheme,
      defaultUnitSystem: backup.settings[Settings.defaultUnitSystem],
      firstDayOfTheWeek,
    });

    if (Platform.OS === 'android') {
      StatusBar.setBackgroundColor(
        appTheme === 'default' ? '#233656' : '#000000'
      );
    }

    realm.write(() => {
      realm.deleteAll();

      backup.exercises.forEach(exercise => {
        realm.create(EXERCISE_SCHEMA_NAME, exercise);
      });

      backup.workouts.forEach(workout => {
        realm.create(WORKOUT_SCHEMA_NAME, workout);
      });
    });
  } else {
    // TODO warn not supported file
  }
};
