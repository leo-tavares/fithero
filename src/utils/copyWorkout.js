/* @flow */

import type {
  AddWorkoutSchemaType,
  WorkoutSchemaType,
} from '../database/types';
import { toDate } from './date';
import realm from '../database';
import { WORKOUT_SCHEMA_NAME } from '../database/schemas/WorkoutSchema';
import {
  deserializeWorkout,
  getExerciseSchemaId,
  getNextSetIndex,
  getSetSchemaId,
} from '../database/utils';
import { getWorkoutByPrimaryKey } from '../database/services/WorkoutService';

export const copyWorkout = (
  selectedWorkout: WorkoutSchemaType,
  selectedDay: string
) => {
  const existingWorkout = getWorkoutByPrimaryKey(selectedDay);

  if (existingWorkout) {
    mergeWorkout(selectedDay, existingWorkout, selectedWorkout);
    return;
  }

  const toCopyWorkout = deserializeWorkout(selectedWorkout);
  const date = toDate(selectedDay);

  const newWorkout: AddWorkoutSchemaType = {
    id: selectedDay,
    date,
    // comments: '', // Do not copy comments
    exercises: [],
  };

  toCopyWorkout.exercises.forEach(e => {
    const exerciseIdDb = getExerciseSchemaId(selectedDay, e.type);
    const sets = e.sets.map((s, index) => {
      const setId = getSetSchemaId(selectedDay, e.type, index + 1);
      return {
        ...s,
        id: setId,
        date,
      };
    });
    newWorkout.exercises.push({
      ...e,
      id: exerciseIdDb,
      date,
      sets,
    });
  });

  realm.write(() => {
    realm.create(WORKOUT_SCHEMA_NAME, newWorkout);
  });
};

const mergeWorkout = (
  selectedDay,
  existingWorkout: WorkoutSchemaType,
  selectedWorkout: WorkoutSchemaType
) => {
  realm.write(() => {
    const date = toDate(selectedDay);
    selectedWorkout.exercises.forEach(exercise => {
      const foundExercise = existingWorkout.exercises.find(
        e => e.type === exercise.type
      );
      if (foundExercise) {
        const lastIndex = getNextSetIndex(foundExercise);

        exercise.sets.forEach((s, i) => {
          const setId = getSetSchemaId(
            selectedDay,
            foundExercise.type,
            lastIndex + i + 1
          );
          foundExercise.sets.push({
            ...s,
            id: setId,
            date,
          });
        });
      } else {
        const exerciseIdDb = getExerciseSchemaId(selectedDay, exercise.type);
        const sets = exercise.sets.map((s, index) => {
          const setId = getSetSchemaId(selectedDay, exercise.type, index + 1);
          return {
            ...s,
            id: setId,
            date,
          };
        });
        existingWorkout.exercises.push({
          ...exercise,
          id: exerciseIdDb,
          sets,
          date,
        });
      }
    });
    // Fix exercises sorting
    existingWorkout.exercises.forEach((exercise, index) => {
      exercise.sort = index + 1;
    });
  });
};
