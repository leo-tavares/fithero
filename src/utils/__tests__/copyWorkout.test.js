/* @flow */

import cloneDeep from 'lodash/cloneDeep';

import { copyWorkout } from '../copyWorkout';
import realm from '../../database';
import { getWorkoutByPrimaryKey } from '../../database/services/WorkoutService';

jest.mock('../date', () => {
  const actualDate = jest.requireActual('../date');
  return {
    ...actualDate,
    toDate: () => new Date('2019-12-28T00:00:00.000Z'),
  };
});

jest.mock('../../database/services/WorkoutService');

describe('copyWorkout', () => {
  const date = new Date('2018-05-04T00:00:00.000Z');
  const workout = {
    id: '20180504',
    date,
    exercises: [
      {
        id: `20180504_bench-press`,
        date,
        type: 'bench-press',
        sort: 1,
        isValid: () => true,
        weight_unit: 'metric',
        sets: [
          {
            id: `20180504_bench-press_001`,
            reps: 5,
            weight: 100,
            date,
            type: 'bench-press',
          },
        ],
      },
      {
        id: `20180504_barbell-squat`,
        date,
        type: 'barbell-squat',
        sort: 2,
        isValid: () => true,
        weight_unit: 'metric',
        sets: [
          {
            id: `20180504_barbell-squat_004`,
            reps: 5,
            weight: 120,
            date,
            type: 'barbell-squat',
          },
          {
            id: `20180504_barbell-squat_001`,
            reps: 4,
            weight: 120,
            date,
            type: 'barbell-squat',
          },
        ],
      },
    ],
    isValid: () => true,
  };

  const existingDate = new Date('2019-12-28T00:00:00.000Z');
  const existingWorkout = {
    id: '20191228',
    date: existingDate,
    exercises: [
      {
        id: `20191228_barbell-squat`,
        date: existingDate,
        type: 'barbell-squat',
        sort: 1,
        isValid: () => true,
        weight_unit: 'metric',
        sets: [
          {
            id: `20191228_barbell-squat_003`,
            reps: 3,
            weight: 150,
            date: existingDate,
            type: 'barbell-squat',
          },
          {
            id: `20191228_barbell-squat_001`,
            reps: 2,
            weight: 150,
            date: existingDate,
            type: 'barbell-squat',
          },
        ],
      },
    ],
    isValid: () => true,
  };

  it('copies a workout into an empty day', () => {
    copyWorkout(workout, '20191228');
    // $FlowFixMe
    expect(workout).toMatchDiffSnapshot(realm.create.mock.calls[0][1]);
  });

  it('merges a previous workout into a non-empty day', () => {
    const realmExistingWorkout = cloneDeep(existingWorkout);
    // $FlowFixMe
    getWorkoutByPrimaryKey.mockImplementationOnce(() => realmExistingWorkout);
    copyWorkout(workout, '20191228');

    // $FlowFixMe
    expect(existingWorkout).toMatchDiffSnapshot(realmExistingWorkout);
    expect(realmExistingWorkout.exercises[0].sets).toMatchSnapshot();
  });
});
