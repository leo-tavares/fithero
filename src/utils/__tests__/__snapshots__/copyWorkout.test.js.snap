// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copyWorkout copies a workout into an empty day 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
-   "date": 2018-05-04T00:00:00.000Z,
+   "date": 2019-12-28T00:00:00.000Z,
    "exercises": Array [
      Object {
-       "date": 2018-05-04T00:00:00.000Z,
-       "id": "20180504_bench-press",
-       "isValid": [Function isValid],
+       "date": 2019-12-28T00:00:00.000Z,
+       "id": "20191228_bench-press",
        "sets": Array [
          Object {
-           "date": 2018-05-04T00:00:00.000Z,
-           "id": "20180504_bench-press_001",
+           "date": 2019-12-28T00:00:00.000Z,
+           "id": "20191228_bench-press_001",
            "reps": 5,
            "type": "bench-press",
            "weight": 100,
          },
        ],
        "sort": 1,
        "type": "bench-press",
        "weight_unit": "metric",
      },
      Object {
-       "date": 2018-05-04T00:00:00.000Z,
-       "id": "20180504_barbell-squat",
-       "isValid": [Function isValid],
+       "date": 2019-12-28T00:00:00.000Z,
+       "id": "20191228_barbell-squat",
        "sets": Array [
          Object {
-           "date": 2018-05-04T00:00:00.000Z,
-           "id": "20180504_barbell-squat_004",
+           "date": 2019-12-28T00:00:00.000Z,
+           "id": "20191228_barbell-squat_001",
            "reps": 5,
            "type": "barbell-squat",
            "weight": 120,
          },
          Object {
-           "date": 2018-05-04T00:00:00.000Z,
-           "id": "20180504_barbell-squat_001",
+           "date": 2019-12-28T00:00:00.000Z,
+           "id": "20191228_barbell-squat_002",
            "reps": 4,
            "type": "barbell-squat",
            "weight": 120,
          },
        ],
        "sort": 2,
        "type": "barbell-squat",
        "weight_unit": "metric",
      },
    ],
-   "id": "20180504",
-   "isValid": [Function isValid],
+   "id": "20191228",
  }
`;

exports[`copyWorkout merges a previous workout into a non-empty day 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -18,13 +18,44 @@
            "id": "20191228_barbell-squat_001",
            "reps": 2,
            "type": "barbell-squat",
            "weight": 150,
          },
+         Object {
+           "date": 2019-12-28T00:00:00.000Z,
+           "id": "20191228_barbell-squat_004",
+           "reps": 5,
+           "type": "barbell-squat",
+           "weight": 120,
+         },
+         Object {
+           "date": 2019-12-28T00:00:00.000Z,
+           "id": "20191228_barbell-squat_005",
+           "reps": 4,
+           "type": "barbell-squat",
+           "weight": 120,
+         },
        ],
        "sort": 1,
        "type": "barbell-squat",
+       "weight_unit": "metric",
+     },
+     Object {
+       "date": 2019-12-28T00:00:00.000Z,
+       "id": "20191228_bench-press",
+       "isValid": [Function isValid],
+       "sets": Array [
+         Object {
+           "date": 2019-12-28T00:00:00.000Z,
+           "id": "20191228_bench-press_001",
+           "reps": 5,
+           "type": "bench-press",
+           "weight": 100,
+         },
+       ],
+       "sort": 2,
+       "type": "bench-press",
        "weight_unit": "metric",
      },
    ],
    "id": "20191228",
    "isValid": [Function isValid],
`;

exports[`copyWorkout merges a previous workout into a non-empty day 2`] = `
Array [
  Object {
    "date": 2019-12-28T00:00:00.000Z,
    "id": "20191228_barbell-squat_003",
    "reps": 3,
    "type": "barbell-squat",
    "weight": 150,
  },
  Object {
    "date": 2019-12-28T00:00:00.000Z,
    "id": "20191228_barbell-squat_001",
    "reps": 2,
    "type": "barbell-squat",
    "weight": 150,
  },
  Object {
    "date": 2019-12-28T00:00:00.000Z,
    "id": "20191228_barbell-squat_004",
    "reps": 5,
    "type": "barbell-squat",
    "weight": 120,
  },
  Object {
    "date": 2019-12-28T00:00:00.000Z,
    "id": "20191228_barbell-squat_005",
    "reps": 4,
    "type": "barbell-squat",
    "weight": 120,
  },
]
`;
